runtime/vam/expr/function/fields.go

```diff
@@ -0,0 +1,88 @@
+package function
+
+import (
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/vector"
+)
+
+// https://github.com/brimdata/zed/blob/main/docs/language/functions.md#fields
+type Fields struct {
+	zctx     *zed.Context
+	innerTyp *zed.TypeArray
+	outerTyp *zed.TypeArray
+}
+
+func NewFields(zctx *zed.Context) *Fields {
+	inner := zctx.LookupTypeArray(zed.TypeString)
+	return &Fields{
+		zctx:     zctx,
+		innerTyp: inner,
+		outerTyp: zctx.LookupTypeArray(inner),
+	}
+}
+
+func (f *Fields) Call(args ...vector.Any) vector.Any {
+	val := vector.Under(args[0])
+	switch typ := val.Type().(type) {
+	case *zed.TypeRecord:
+		paths := buildPath(typ, nil)
+		s := vector.NewStringEmpty(val.Len(), nil)
+		inOffs, outOffs := []uint32{0}, []uint32{0}
+		for i := uint32(0); i < val.Len(); i++ {
+			inOffs, outOffs = appendPaths(paths, s, inOffs, outOffs)
+		}
+		inner := vector.NewArray(f.innerTyp, inOffs, s, nil)
+		return vector.NewArray(f.outerTyp, outOffs, inner, nil)
+	case *zed.TypeOfType:
+		var errcnt uint32
+		tags := make([]uint32, val.Len())
+		s := vector.NewStringEmpty(val.Len(), nil)
+		inOffs, outOffs := []uint32{0}, []uint32{0}
+		for i := uint32(0); i < val.Len(); i++ {
+			b, _ := vector.TypeValueValue(val, i)
+			rtyp := f.recordType(b)
+			if rtyp == nil {
+				errcnt++
+				tags[i] = 1
+				continue
+			}
+			inOffs, outOffs = appendPaths(buildPath(rtyp, nil), s, inOffs, outOffs)
+		}
+		inner := vector.NewArray(f.innerTyp, inOffs, s, nil)
+		out := vector.NewArray(f.outerTyp, outOffs, inner, nil)
+		return vector.NewVariant(tags, []vector.Any{out, vector.NewStringError(f.zctx, "missing", errcnt)})
```

Comment:
The common case is to not have a variant result here.  I think we should have a check for `errcnt == 0` check and just return `out` in this case...?